icu4c/source/common/locid.cpp

```diff
@@ -1340,7 +1340,25 @@ Locale::getUnicodeKeywordValue(StringPiece keywordName,
 void
 Locale::setKeywordValue(const char* keywordName, const char* keywordValue, UErrorCode &status)
 {
-    uloc_setKeywordValue(keywordName, keywordValue, fullName, ULOC_FULLNAME_CAPACITY, &status);
+    int32_t bufferLength = uprv_max((int32_t)(uprv_strlen(fullName) + 1), ULOC_FULLNAME_CAPACITY);
+    int32_t newLength = uloc_setKeywordValue(keywordName, keywordValue, fullName,
+                                         bufferLength, &status) + 1;
+    /* Handle the case the current buffer is not enough to hold the new id */
+    if (status == U_BUFFER_OVERFLOW_ERROR || newLength > bufferLength) {
```

Comment:
This should be `&&` instead of `||`, shouldn't it?

Unless something inside of `uloc_setKeywordValue()` is broken, it should not be possible for one of these two conditions to be true while the other is not.

If you want to be defensive against such a possible bug, you could add an assertion here to verify that these two conditions are always either both true or both false:

```
    if (status == U_BUFFER_OVERFLOW_ERROR) {
        U_ASSERT(newLength > bufferLength);
        // ...
    } else {
        U_ASSERT(newLength <= bufferLength);
    }
```

But the way the code is written now, with `||` instead of `&&`, it looks as if a value of `status` that is something other than `U_BUFFER_OVERFLOW_ERROR` could possibly be overwritten with `U_ZERO_ERROR` below (which would be wrong do to).

Reply:
see existing code in Locale::init